server/aws-lsp-partiql/src/server/syntax-highlighting/parser-tokens.ts

```diff
@@ -0,0 +1,195 @@
+import Parser from 'web-tree-sitter'
+import parserBase64 from '../tree-sitter-parser/tree-sitter-parser-inline'
+import wasmBinaryArray from '../tree-sitter-parser/tree-sitter-inline'
+import {
+    SemanticTokenTypes,
+    SemanticTokens,
+    SemanticTokenModifiers,
+    Range,
+} from '@aws/language-server-runtimes/server-interface'
+import { SemanticTokensBuilder } from 'vscode-languageserver/node'
+import { semanticTokensLegend } from '../language-service'
+
+export interface SemanticToken {
+    range: Range
+    tokenType: SemanticTokenTypes
+    tokenLength: number
+    /**
+     * An optional array of modifiers for this token
+     */
+    tokenModifiers?: SemanticTokenModifiers[]
+}
+
+export const string2TokenTypes: { [key: string]: SemanticTokenTypes } = {
+    tuple_punc_start: SemanticTokenTypes.operator,
+    tuple_punc_end: SemanticTokenTypes.operator,
+    tuple_punc_separator: SemanticTokenTypes.operator,
+    pair_punc_separator: SemanticTokenTypes.operator,
+    array_punc_start: SemanticTokenTypes.operator,
+    array_punc_end: SemanticTokenTypes.operator,
+    array_punc_separator: SemanticTokenTypes.operator,
+    bag_punc_start: SemanticTokenTypes.operator,
+    bag_punc_end: SemanticTokenTypes.operator,
+    bag_punc_separator: SemanticTokenTypes.operator,
+    ion: SemanticTokenTypes.string,
+}
+
+// Initialize and prepare the parser
+async function initParser() {
+    let wasmModuleUrl: string | null = null
+    if (typeof window !== 'undefined' && typeof window.URL.createObjectURL === 'function') {
+        // Browser
+        const wasmBlob = new Blob([wasmBinaryArray], { type: 'application/wasm' })
+        wasmModuleUrl = URL.createObjectURL(wasmBlob)
+    } else if (typeof process !== 'undefined' && typeof process.versions.node !== 'undefined') {
+        // Node.js
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        const fs = require('fs')
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        const path = require('path')
+        const tempWasmPath = path.join(__dirname, 'temp-tree-sitter.wasm')
+        fs.writeFileSync(tempWasmPath, Buffer.from(wasmBinaryArray))
+        wasmModuleUrl = tempWasmPath
+    }
+    await Parser.init({
+        locateFile(path: string, _prefix: string) {
+            if (path.endsWith('.wasm')) {
+                return wasmModuleUrl
+            }
+            return path
+        },
+    })
+    const parser = new Parser()
+    const PartiQL = await Parser.Language.load(parserBase64)
+    parser.setLanguage(PartiQL)
+    return parser
+}
+
+export async function findNodes(
+    sourceCode: string | Parser.Input,
+    nodeType: SemanticTokenTypes | string
+): Promise<SemanticToken[]> {
+    const parser = await initParser()
```

Comment:
We now initialize the parser every time the client requests semantic tokens, that's probably a bit inefficient. Could we run `initParser` only once when the server/service gets created and then reuse it?

Reply:
You are right. Current implementation is quite inefficient and created a global parser for reusage in a latest commit.